jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHttpOutputInterceptor.java

```diff
@@ -318,54 +328,69 @@ protected Action process() throws Exception
             Deflater deflater = _deflaterEntry.get();
             if (!deflater.finished())
             {
-                if (deflater.needsInput())
+                if (_channel.isUseOutputDirectByteBuffers())
                 {
-                    // if there is no more content available to compress
-                    // then we are either finished all content or just the current write.
-                    if (BufferUtil.isEmpty(_content))
+                    if (deflater.needsInput() && !_last)
                     {
-                        if (_last)
-                            deflater.finish();
-                        else
-                            return Action.SUCCEEDED;
+                        return Action.SUCCEEDED;
                     }
-                    else
+
+                    int pos = BufferUtil.flipToFill(_buffer);
+                    int produced = deflater.deflate(_buffer, _syncFlush ? Deflater.SYNC_FLUSH : Deflater.NO_FLUSH);
+                    BufferUtil.flipToFlush(_buffer, pos);
+                }
+                else
+                {
```

Comment:
I would like to see us remove this branch completely and always use the `ByteBuffer` APIs for `Deflater` and `CRC32`.

If the buffer is not direct and does not have an array then `CRC32.update` will not copy the entire buffer at once, but allocate a single byte array of max size 4k and use that to read from the buffer.

Reply:
Done!